### DISCUSSION TITLE
Can someone clarify processing of CKA_NSS_SERVER_DISTRUST_AFTER by curl

### DISCUSSION THREAD

**Discussion by artem-kamyshev** [curl/curl#10058] [Q&A] [answered]
Hello there,

I've been searching info about CKA_NSS_SERVER_DISTRUST_AFTER usage in certificates provided by Mozilla and found out this issue: https://github.com/curl/curl/issues/4834

Here is the commit: https://github.com/curl/curl/commit/1ebc53df251819086859f27f22ccfb66ddd58364

I'm not a curl or perl expert so I kind ask for your help: please clarify if I go everything right.

Curl uses these certificates when making TLS connections. When there is a field CKA_NSS_SERVER_DISTRUST_AFTER in certificates got from Mozilla curl applies the following logic to them:
1. If current time is less then set by CKA_NSS_SERVER_DISTRUST_AFTER this certificate is trusted is added to a curl's bundle of CA's certificates.
2. If current time is equal or greater then set by CKA_NSS_SERVER_DISTRUST_AFTER this certificate is not trusted anymore and is not added to curl's bundle of CA's certificates. **Thus it is never used in certificate verification chain and all other certificates which are signed by it are not trusted anymore.**

Did I get it right?

**Comment by bagder** [comment, marked as answer]
`CKA_NSS_SERVER_DISTRUST_AFTER` is a parameter that might be set on a cert in the Mozilla CA bundle. It is a 13 byte time stamp.

When `mk-ca-bundle.pl` is run and converts the Mozilla CA cert file into a PEM version, it recognizes this parameter and will exclude that certificate from the output if the time stamp has passed. The timestamp itself only resides in the Mozilla file and not in the final PEM file, as there is no way to provide such meta data there.

If you create the PEM the file a day before before the trust runs out, and then use the PEM the following month, it will work fine with curl and all tools that use this PEM file and they will not be aware of the end of trust date. Only when `mk-ca-bundle.pl` is rerun on the Mozilla file to recreate the PEM file will the expired trust have an effect and then exclude that particular cert from the output.

The [CA extract service](https://curl.se/docs/caextract.html) on the curl website reruns this script daily and checks if the output is different than previously, so it will generate a new PEM the day the trust runs out for a certificate in the bundle.

When using a PEM bundle where a CA certificate is no longer included, curl fails to verify certificates that were signed with a CA that is no longer trusted since the bundle no longer contains  that CA's certificate.
